server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterHealthIndicatorService.java

```diff
@@ -409,28 +425,89 @@ private HealthIndicatorResult getMasterIsStableResult(boolean explain, MasterHis
+            result = getIndicatorResultOnCannotJoinLeader(localMasterHistory, currentMaster, explain);
+        } else {
+            // NOTE: The logic in this block will be implemented in a future PR
+            result = createIndicator(
+                HealthStatus.RED,
+                "No master has been observed recently",
+                HealthIndicatorDetails.EMPTY,
+                UNSTABLE_MASTER_IMPACTS,
+                getContactSupportUserActions(explain)
+            );
+        }
+        return result;
+    }
+
+    /**
+     * Creates a HealthIndicatorResult in the case that there has been no master in the last few seconds, there is no elected master known,
+     * and there are no master eligible nodes. The status will be RED, and the details (if explain is true) will contain the list of any
+     * masters seen previously and a description of known problems from this node's Coordinator.
+     * @param localMasterHistory Used to pull recent master nodes for the details if explain is true
+     * @param explain If true, details are returned
+     * @return A HealthIndicatorResult with a RED status
+     */
+    private HealthIndicatorResult getIndicatorResultOnNoMasterEligibleNodes(MasterHistory localMasterHistory, boolean explain) {
+        String summary = "No master eligible nodes found in the cluster";
+        HealthIndicatorDetails details = getDetails(explain, localMasterHistory, coordinator.getClusterFormationState().getDescription());
+        return createIndicator(HealthStatus.RED, summary, details, UNSTABLE_MASTER_IMPACTS, getContactSupportUserActions(explain));
+    }
+
+    /**
+     * Creates a HealthIndicatorResult in the case that there has been no master in the last few seconds in this node's cluster state, but
+     * PeerFinder reports that there is an elected master. The assumption is that this node is having a problem joining the elected master.
+     * The status will be RED, and the details (if explain is true) will contain the list of any masters seen previously and a
+     * description of known problems from this node's Coordinator.
+     * @param localMasterHistory Used to pull recent master nodes for the details if explain is true
+     * @param currentMaster The node that PeerFinder reports as the elected master
+     * @param explain If true, details are returned
+     * @return A HealthIndicatorResult with a RED status
+     */
+    private HealthIndicatorResult getIndicatorResultOnCannotJoinLeader(
+        MasterHistory localMasterHistory,
+        DiscoveryNode currentMaster,
+        boolean explain
+    ) {
+        String summary = String.format(
+            Locale.ROOT,
+            "%s has been elected master, but the node being queried, %s, is unable to join it",
+            currentMaster,
+            clusterService.localNode()
         );
+        HealthIndicatorDetails details = getDetails(explain, localMasterHistory, coordinator.getClusterFormationState().getDescription());
```

Comment:
Similarly, do we want an established structure in the api details? ie. display `ClusterFormationState#inFlightJoinStatuses` in a structured way?